lib/server.js

```diff
@@ -117,15 +119,47 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
     routes.notify, brRest.when.prefers.ld,
     validate('continuity-server.notification'),
     asyncHandler(async (req, res) => {
-      // FIXME: we need to authenticate `peerId`; http-signatures should be
-      // added here
+      const verified = await _verifyRequest({req});
+      if(!verified) {
+        return res.status(403).end();
+      }
       // queue handling notification and immediately send response, do not
       // wait for notification handling to complete
       _handleNotification({req}).catch(_logNotificationError);
       res.status(204).end();
     }));
 });
 
+async function _verifyRequest({req}) {
+  try {
+    const {url, method, headers} = req;
+    const expectedHeaders = [
+      '(created)',
+      '(expires)',
+      // (key-id) will include the peerId in the signature itself
+      '(key-id)',
+      '(request-target)',
+      'digest',
+      'content-type',
+      'host'
+    ];
+    const parsed = parseRequest(
+      {url, method, headers}, {headers: expectedHeaders});
+    // get parsed parameters from from HTTP header and generate signing string
+    const {keyId, signingString, params: {signature: b64Signature}} = parsed;
+    const key = await getKeyPair({peerId: keyId});
+    const verifier = key.verifier();
+    const encoder = new TextEncoder();
+    const data = encoder.encode(signingString);
+    const signature = Buffer.from(b64Signature, 'base64');
+    // verify HTTP signature
+    return verifier.verify({data, signature});
+  } catch(e) {
+    logger.error('_verifyRequest errored', {error: e});
+    return false;
```

Comment:
I think we want to surface errors that prevented verification from even running (vs. just a bad verify) to the client, e.g., if their signature is in the wrong format.

The previous code would have returned a 500 if their signature format was bad, which we don't want (implying the server had an issue, which it didn't). But this new code will hide too much. What we want is, if the client data is bad/the mistake is on them, we want to return an appropriate error code and error message.

Reply:
first attempt here: https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/pull/249/commits/9bad307ee998b689b905f616b4a3c18eefb88db2